ext/telemetry/lib.rs

```diff
@@ -974,188 +1057,545 @@ fn owned_string<'s>(
+    let span_context = match &**state {
+      OtelSpanState::Recording(span) => &span.span_context,
+      OtelSpanState::Done(span_context) => span_context,
+    };
+    JsSpanContext {
+      trace_id: format!("{:?}", span_context.trace_id()).into(),
+      span_id: format!("{:?}", span_context.span_id()).into(),
+      trace_flags: span_context.trace_flags().to_u8(),
+    }
+  }
+
+  #[fast]
+  fn set_status<'s>(
+    &self,
+    #[smi] status: u8,
+    #[string] error_description: String,
+  ) -> Result<(), AnyError> {
+    let mut state = self.0.borrow_mut();
+    let OtelSpanState::Recording(span) = &mut **state else {
+      return Ok(());
+    };
+    span.status = match status {
       0 => SpanStatus::Unset,
       1 => SpanStatus::Ok,
       2 => SpanStatus::Error {
-        description: Cow::Owned(error_description.into_owned()),
+        description: Cow::Owned(error_description),
       },
-      _ => return,
+      _ => return Err(type_error("invalid span status code")),
+    };
+    Ok(())
+  }
+
+  #[fast]
+  fn drop_event(&self) {
+    let mut state = self.0.borrow_mut();
+    match &mut **state {
+      OtelSpanState::Recording(span) => {
+        span.events.dropped_count += 1;
+      }
+      OtelSpanState::Done(_) => {}
+    }
+  }
+
+  #[fast]
+  fn drop_link(&self) {
+    let mut state = self.0.borrow_mut();
+    match &mut **state {
+      OtelSpanState::Recording(span) => {
+        span.links.dropped_count += 1;
+      }
+      OtelSpanState::Done(_) => {}
+    }
+  }
+
+  #[fast]
+  fn end(&self, end_time: f64) {
+    let end_time = if end_time.is_nan() {
+      SystemTime::now()
+    } else {
+      SystemTime::UNIX_EPOCH
+        .checked_add(Duration::from_secs_f64(end_time))
+        .unwrap()
     };
+
+    let mut state = self.0.borrow_mut();
+    if let OtelSpanState::Recording(span) = &mut **state {
+      let span_context = span.span_context.clone();
+      if let OtelSpanState::Recording(mut span) = *std::mem::replace(
+        &mut *state,
+        Box::new(OtelSpanState::Done(span_context)),
+      ) {
+        span.end_time = end_time;
+        let Some(OtelGlobals { span_processor, .. }) = OTEL_GLOBALS.get()
+        else {
+          return;
+        };
+        span_processor.on_end(span);
+      }
+    }
   }
 }
 
 #[op2(fast)]
-fn op_otel_span_attribute<'s>(
+fn op_otel_span_attribute1<'s>(
   scope: &mut v8::HandleScope<'s>,
-  state: &mut OpState,
-  #[smi] capacity: u32,
+  span: v8::Local<'_, v8::Value>,
   key: v8::Local<'s, v8::Value>,
   value: v8::Local<'s, v8::Value>,
 ) {
-  if let Some(temporary_span) = state.try_borrow_mut::<TemporarySpan>() {
-    temporary_span.0.attributes.reserve_exact(
-      (capacity as usize)
-        .saturating_sub(temporary_span.0.attributes.capacity()),
-    );
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key, value);
+  let Some(span) =
+    deno_core::_ops::try_unwrap_cppgc_object::<OtelSpan>(scope, span)
```

Comment:
@littledivy I think we should reexport this function in `deno_core` so one doesn't have to access `_ops`. It's not visible in https://docs.rs/deno_core/latest/deno_core/ either.

Reply:
Consider adding support for `let span = cppgc!(OtelSpan, scope, span)`, as this `let ... else ...` pattern seems to be often used.

Reply:
I just have to use this due to another bug in core. I can clean this up soon.